### PR TITLE
 add similar to the z-index function of CSS to positioned

### PR DESCRIPTION
**Pull request by northka**
## Description

 add similar to the z-index function of CSS to positioned

## Related Issues

[50763](https://github.com/flutter/flutter/issues/50763)
## Tests

I added the following tests:
- Stack paint child in order according to child zIndex

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by northka** [comment]
@googlebot I signed it!

File: packages/flutter/lib/src/rendering/stack.dart [hunk 1]
```diff
@@ -595,11 +599,34 @@ class RenderStack extends RenderBox
 
   /// Override in subclasses to customize how the stack paints.
   ///
-  /// By default, the stack uses [defaultPaint]. This function is called by
-  /// [paint] after potentially applying a clip to contain visual overflow.
+  /// This function is called by [paint] after potentially applying
+  /// a clip to contain visual overflow.
+  /// children are sorted by child's zIndex in an ascending order then painted
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/stack.dart:604]: nit: doc comments should be full and grammatically correct sentences starting with a capital letter and ending with a `.`.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -3585,6 +3597,11 @@ class Positioned extends ParentDataWidget<StackParentData> {
       needsLayout = true;
     }
 
+    if (parentData.zIndex!= zIndex) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3600]:
```suggestion
    if (parentData.zIndex != zIndex) {
```

File: packages/flutter/lib/src/widgets/basic.dart [hunk 3]
```diff
@@ -3549,6 +3555,12 @@ class Positioned extends ParentDataWidget<StackParentData> {
   /// vertically.
   final double height;
 
+  /// The child's Z-index
+  ///
+  /// This integer is the stack level of the child in the current stacking
+  ///  context.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3561]: nit: remove leading space.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 4]
```diff
@@ -3549,6 +3555,12 @@ class Positioned extends ParentDataWidget<StackParentData> {
   /// vertically.
   final double height;
 
+  /// The child's Z-index
+  ///
+  /// This integer is the stack level of the child in the current stacking
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3560]: This doc is not very clear.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 5]
```diff
@@ -3705,6 +3724,12 @@ class PositionedDirectional extends StatelessWidget {
   /// {@macro flutter.widgets.child}
   final Widget child;
 
+  /// The child's Z-index
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/basic.dart:3727]: Same here.

**Comment by goderbauer** [comment]
The Stack is designed and documented to paint children in the order they appear in the children list. I don't think we want to add another sorting mechanism to this. z-index seems especially problematic since it lends itself to z fighting.

**Pull request closed without merging**